### PR TITLE
Lookup tables (binary search)

### PR DESCRIPTION
**Pull request by DL6ER**
# What does this implement/fix?

So far, `pihole-FTL` uses linear searching (scanning through an array) to find entries in the tables holding
- domains,
- clients, and
- DNS cache data.

Even though we have added quite some enhancements to this (e.g. matching by a precomputed hash instead of comparing lengthy strings), a linear search is still an algorithm that needs linear time to find items. This is not actually a bad strategy if there isn't much to be searched as all other alternatives comes their own overhead.

This PR replaces the legacy linear searching by a customized binary search algorithm to *vastly* improve the performance of these lookups. Binary searching is a more efficient algorithm for searching sorted arrays, requiring only logarithmic time to find an item. This significant reduction in the number of steps needed for a search operation can lead to substantial performance improvements, especially as the size of the data grows.

## Performance Comparison:
- **Linear Search**
  - Time Complexity: In the worst case, linear search requires examining each element in the array, resulting in $\mathcal{O}(N)$ time complexity.
  - Use Case: Linear search is simple and effective for small datasets where the overhead of sorting and maintaining a sorted array may not be justified.
  - Drawbacks: As the size of the dataset increases, the time required for linear search grows linearly, making it less efficient for large datasets.

- **Binary Search**
  - Time Complexity: Binary search operates on sorted arrays and divides the search interval in half with each step, resulting in $\mathcal{O}(\log N)$ time complexity.
  - Use Case: Binary search is highly efficient for large datasets where the overhead of maintaining a sorted array is outweighed by the performance gains during search operations.
  - Advantages: The logarithmic time complexity of binary search means that even for large datasets, the number of steps required to find an item remains relatively small.
  - Drawbacks: Binary search requires the array to be sorted, which introduces additional complexity and overhead for maintaining the sorted order. We ensure the array always stays sorted by inserting new entries at the location where a sorting algorithm would put them.

**Okay, enough theory, give me some *concrete* figures!**

Sure, here are some numbers - they are actual measurements of the code in this PR with the optional compile-time flag `MEASURE_ITERATIONS` enabled:

- before 125, now 9 - $\log_2(125) \approx 6.9$
- before 336, now 10 - $\log_2(336) \approx 8.4$
- before 991, now 11 - $\log_2(991) \approx 9.9$
- before 1111, now 11 - $\log_2(1111) \approx 10.1$
- before 29868, now 16 - $\log_2(29868) \approx 14.9$

Why is the actual number of iterations sometimes hogher than the theoretical prediction? Well, our implementation may sometimes need a few extra steps when hash collisions occur (we use a very simple = fast hash) as we then have to do some mini-form of linear searching adding up to a few extra iterations.

## Implementation Details:
During a search, the binary search algorithm repeatedly divides the search interval in half, comparing the target value to the middle element of the interval. Depending on the comparison result, the search interval is adjusted to either the left or right half, effectively narrowing down the search space by 50% during each iteration.

Performance Gains: The switch from linear to binary search has resulted in noticeable performance improvements, particularly for large datasets. The reduced number of comparisons and steps required to find an item translates to faster lookup times and more efficient resource utilization.

In summary, while linear search was a straightforward and effective strategy, the implementation of binary search significantly enhances the performance of `pihole-FTL` especially for active network. The logarithmic time complexity of binary search ensures that lookup operations remain efficient, even as the size of the data grows, providing a more scalable solution for managing domains, clients, and DNS cache data on Internet Service Provider (ISP) scales.

---

**Related issue or feature (if applicable):** N/A

**Pull request in [docs](https://github.com/pi-hole/docs) with documentation (if applicable):** N/A

---
**By submitting this pull request, I confirm the following:** 

1. I have read and understood the [contributors guide](https://docs.pi-hole.net/guides/github/contributing/), as well as this entire template. I understand which branch to base my commits and Pull Requests against. 
2. I have commented my proposed changes within the code.
3. I am willing to help maintain this change if there are issues with it later.
4. It is compatible with the [EUPL 1.2 license](https://opensource.org/licenses/EUPL-1.1)
5. I have squashed any insignificant commits. ([`git rebase`](http://gitready.com/advanced/2009/02/10/squashing-commits-with-rebase.html))

## Checklist:

- [x] The code change is tested and works locally.
- [x] I based my code and PRs against the repositories `developmental` branch.
- [x] I [signed off](https://docs.pi-hole.net/guides/github/how-to-signoff/) all commits. Pi-hole enforces the [DCO](https://docs.pi-hole.net/guides/github/dco/) for all contributions
- [x] I [signed](https://docs.github.com/en/authentication/managing-commit-signature-verification/signing-commits) all my commits. Pi-hole requires signatures to verify authorship
- [x] I have read the above and my PR is ready for review.

### CODE REVIEW

**Comment by yubiuser** [comment]
>  for searching sorted arrays

> DNS cache data

Doesn't this mean, FTL need to re-sort the array every time new cache data is added/removed (e.g. new query arrives or old is removed from cache)?

**Comment by DL6ER** [comment]
> Doesn't this mean, FTL need to re-sort the array every time new cache data is added/removed (e.g. new query arrives or old is removed from cache)?

No, see above: 

>  [...] We ensure the array always stays sorted by inserting new entries at the location where a sorting algorithm would put them. [...]

This is the main justification for not using the C standard function `bsearch()`, cfm. [`man bsearch`](https://man7.org/linux/man-pages/man3/bsearch.3.html):

> **RETURN VALUE**
>
>   The `bsearch()` function returns a pointer to a matching member of
>   the array, **or NULL if no match is found**.  If there are multiple
>   elements that match the key, the element returned is unspecified.

Returning `NULL` in the case of not-found is unfortunate here as the algorithm already looked at the exact position where the hash *should* have been. Our own implementation returns the pointer even in the not-found case indicating the not-found property by an additional boolean. Now that we know the "where it should have been" location, we can right away do two things:
1. Move all entries from this location one slot to the right (all shared memory objects always have room for at least one item at the end). This is done at once using `memmove()` and not using a loop. This is very fast.
2. Insert the new item exactly at this location

Hence, we do not ever need to sort the array.

Example: Assume we have the array `1 2 3 4 6 7 8 9 X` (X = empty) and are searching for `5`. We determine that it is not there. The standard implementation would return `NULL` and we'd have to manually search the array (time-consuming). Our implementation returns the pointer to `4` and `false`. We now do the two steps above to get:
1. `1 2 3 4 X 6 7 8 9` (everything *behind* 4 is moved one up) 
2. `1 2 3 4 5 6 7 8 9` (`5` is inserted)

We use a similar trick when removing an item from the lookup table. We use the known ID and hash at that time and use our function to rapidly find the exact entry. We then take everything that is right of this and overwrite the entry to be removed. If we want to remove the last, there is nothing to move and we zero out the entry, instead.

As we *always* do exactly apply adding/removing in-place at the exact right position from the beginning (lookup tables completely empty), there is never a need to do any kind of sorting elegantly avoiding any "real" sorting overhead.

TL;DR: The trick really is that our search function *knows* where to add this so adding is basically no work. Removing can trivially use the searching to find the element to search. Everything else is just memory moving which is blazing fast. Mind also that lookup tables will be ~ KB, they take 8 bytes per entry on a 64bit machine. Even millions of entries will be on the ~ low MB range, which is basically nothing concerning today's memory speeds.

**Comment by DL6ER** [comment]
@sawsanders Thanks for helping reviewing this PR and I am delighted to hear you are seeing some speedup in daily use! 

Changing the integer type from signed to unsigned already brought up a few oversights I have fixed previously in this PR. The one you found is another one but (fortunately) entirely harmless: there is protecting code in place that did the right thing here - plus printing the warning I have now addressed in f64ce8f.

File: src/api/theme.h [hunk 1]
```diff
@@ -24,11 +24,11 @@ enum web_theme {
 } __attribute__ ((packed));
 
 struct web_themes{
-	const enum web_theme id;
 	const char *name;
 	const char *description;
-	const bool dark;
 	const char *color;
+	const enum web_theme id;
+	const bool dark;
```
Comment (DL6ER) [inline src/api/theme.h:31]: Rearranging `struct` content so save space

**Comment by sawsanders** [comment]
The errors in the FTL log are gone. 👍 

I am seeing domains in the query log that are forwarded and then subsequently blocked (the underlying domain is in my blocklists and should be blocked). It appears to be related to deep CNAME inspection. I don't remember seeing this behavior before but maybe it's always been that way and I'm just now noticing.

Example:
qlog

pihole.log:
pihole log

If normal and expected, please disregard. Thanks!

**Comment by DL6ER** [comment]
@sawsanders Thanks for confirming. Looking at your screenshot, this seems to be fine as deep CNAME inspection is indeed supposed to be intervening here if you block `mailgun.org`, as 
```
email.comms.awb.org.    300     IN      CNAME   mailgun.org.
mailgun.org.            300     IN      A       34.110.180.34
```

File: src/lookup-table.c [hunk 2]
```diff
@@ -0,0 +1,450 @@
...
+ *
+ * This function performs a binary search on an array of lookup_table structures.
+ * It compares the key with the middle element of the array and decides whether
+ * to search in the left or right half of the array. The search continues until
+ * the element is found or the array is exhausted. This function is used to find
+ * an element in a sorted array of lookup_table structures based on the hash value.
+ * Finding the element in the array is done in O(log n) time complexity.
+ *
+ * This function is different from the standard bsearch function in that it
+ * always the address of where the element would be found if it were in the
```
Comment (yubiuser) [inline src/lookup-table.c:66]:
```suggestion
 * always returns the address of where the element would be found if it were in the
```

File: src/lookup-table.c [hunk 3]
```diff
@@ -0,0 +1,450 @@
...
+		// Compare the key with the current element
+		const int sign = cmp_hash(hash, (*try)->hash);
+
+		if(sign == 0)
+		{
+			// If the key matches the element, we found at least one
+			// occurrence of the key in the array
+			return true;
+		}
+		else if(nel == 1)
```
Comment (yubiuser) [inline src/lookup-table.c:106]: From a reading perspective I would have expected a check for `sign <0` here and only within this case a check for `nel == 1` and not the other way round

**Comment by yubiuser** [comment]
> Step 3
Now consider the second remaining array. The middle element is 13, which matches the target value. 

How can there be a middle if a even number of elements exist? Does it always picks the bigger one? Or is this random?

File: src/api/theme.h (hunk 1 above)
Comment (DL6ER) [reply src/api/theme.h:31]: Maybe of interest here: The "old" way takes 40 bytes of total size on 64 bit (20 bytes on 32 bit). Thew new ordering is 32 bytes on 64 bit, unchanged 20 bytes on 32 bit systems.

Comment (yubiuser) [reply src/api/theme.h:31]: Thanks!

File: src/lookup-table.c (hunk 3 above)
Comment (DL6ER) [reply src/lookup-table.c:106]: but we can have `nel == 1` and `sign > 0` in which case we have the code
``` c
			// If the key is greater than the middle element, the key
			// would be inserted after the middle element
			(*try)++;
```

**Comment by DL6ER** [comment]
> How can there be a middle if a even number of elements exist? Does it always picks the bigger one? Or is this random?

It is `idx = size/2`. In our case, `size == 2` (of the remaining array, so we have to look at index 1. If the size is 1, integer division (!) gives us `1/2 == 0`. `3/2 = 1`, `4/2 = 2`, etc. It is deterministic.

File: src/lookup-table.c [hunk 4]
```diff
@@ -0,0 +1,450 @@
...
+	          type, id, hash, pos);
+
+	// Increase the number of elements in the array
+	(*size)++;
+
+	return true;
+}
+
+/**
+ * @brief Inserts an element into the lookup table.
```
Comment (yubiuser) [inline src/lookup-table.c:240]:
```suggestion
 * @brief Removes an element from the lookup table.
```

**Review by yubiuser — COMMENTED** [review]
I saw those debug lines 

```
2024-10-16 21:33:00.367 CEST [286460/T286462] DEBUG_GC: Element to be found (type 7, hash 1170501315) is not in the lookup table
2024-10-16 21:33:00.368 CEST [286460/T286462] DEBUG_GC: Element to be found (type 7, hash 1915492500) is not in the lookup table
2024-10-16 21:33:00.368 CEST [286460/T286462] DEBUG_GC: Element to be found (type 7, hash 319129926) is not in the lookup table

...

2024-10-16 21:33:14.498 CEST [286460M] DEBUG_GC: Found element (type 7, ID 2, hash 3673561384) at position 7
2024-10-16 21:33:14.500 CEST [286460M] DEBUG_GC: Found element (type 8, ID 51, hash 3165978793) at position 779
2024-10-16 21:33:14.500 CEST [286460M] DEBUG_GC: Found element (type 9, ID 181, hash 3341061574) at position 2026
2024-10-16 21:33:14.501 CEST [286460M] DEBUG_GC: Found element (type 9, ID 181, hash 3341061574) at position 2026
```

Do you think, the `type` should be human readable? Or how do I know that `type 7` corresponds to?
Should the debug tell which `lookup` table was searched (clients, domains,...)?
Also the last lines do not tell us in which table the element was found.

___

The function `print_lookup_table` seems unused.

**Comment by DL6ER** [comment]
> Do you think, the `type` should be human readable? Or how do I know that `type 7` corresponds to?

Yes, I added this in the most recent commit. Just FYI, the types are of type `enum memory_type` and we always have
- `CLIENTS_LOOKUP == 7`
- `DOMAINS_LOOKUP == 8`
- `DNS_CACHE_LOOKUP == 9`

>  Should the debug tell which lookup table was searched (`clients`, `domains`,...)?

Yes, has been added.

>  Also the last lines do not tell us in which table the element was found.

It says it in the "type". The new debug format improved on this.

> The function `print_lookup_table` seems unused.

Yes, a debugging leftover. It is harmless to keep it but I removed it now.

**Comment by DL6ER** [comment]
The last commits implement a debugging tool useful for searching for hash collisions. We can discuss if we leave it in or if we want to remove it before finally merging this PR. But more importantly, it helped me realize (and fix!) that our previous implementation of `hashThreeInts()`.

Our code can handle hash collisions well, but they still mean extra steps to be taken. Hence, we'd like to avoid them if possible. On my testing Pi-hole (20 clients, 2078 domains, 5138 DNS cache entries), I have seen no hash collisions for clients and domains (hash computed in `hashStr()`) but 16 collisions in the DNS cache (hash computed by `hashThreeInts()`). The implementation of the latter hashing function was a mere try without any real tests, so far.

I added an altogether new, *much simpler*, implementation which is easy to understand (see comments in code). With this, I immediately got down to zero hash collisions in all three lookup tables.

Anyway, the sub-optimal previous implementation of `hashThreeInts()` proved that our handling of hash collisions worked in all the cases that occurred.

**Comment by DL6ER** [comment]
Manual inspection of the lookup tables revealed that removing leaves some traces in memory, cfm. `0540 0000 7bf8 ffff` below:
```
[...]
0003730 0801 0000 e338 fda3 04ee 0000 afd0 fdae
0003740 083e 0000 e799 fdc0 06c4 0000 a2e2 fdd4
0003750 00df 0000 777c fe39 0153 0000 f5a9 fe39
0003760 0428 0000 860f fe6b 0317 0000 491b fef0
0003770 06bc 0000 c433 fef5 01ae 0000 0245 ff42
0003780 0139 0000 1345 ff4f 04ea 0000 963e ff6a
0003790 0581 0000 4156 ff91 0019 0000 967c ffd8
00037a0 0540 0000 7bf8 ffff 0540 0000 7bf8 ffff
* <--- this indicates that the previous data is repeated (0x43e0 - 0x37a0)/16 = 197 times
00043e0 0540 0000 7bf8 ffff 0000 0000 0000 0000
00043f0 0000 0000 0000 0000 0000 0000 0000 0000
*
0006000
```
This is *neither* a security nor any other kind of issue as FTL always knows how long the table is but still not "nice". Fixed in most recent commit, now we have:
```
[...]
0003730 0801 0000 e338 fda3 04ee 0000 afd0 fdae
0003740 083e 0000 e799 fdc0 06c4 0000 a2e2 fdd4
0003750 00df 0000 777c fe39 0153 0000 f5a9 fe39
0003760 0428 0000 860f fe6b 0317 0000 491b fef0
0003770 06bc 0000 c433 fef5 01ae 0000 0245 ff42
0003780 0139 0000 1345 ff4f 04ea 0000 963e ff6a
0003790 0581 0000 4156 ff91 0019 0000 967c ffd8
00037a0 0540 0000 7bf8 ffff 0000 0000 0000 0000
*
0006000
```

**Comment by yubiuser** [comment]
Is it expected to recycle the domains that fast?

```
2024-10-19 14:48:24.094 CEST [353469M] DEBUG_GC: New domain: gateway.prod.eu-west-1.forester.a2z.com (ID 1395)
2024-10-19 14:48:24.095 CEST [353469M] DEBUG_GC: New cache entry: domainID 1395, clientID 7, query_type 1 (ID 3577)
2024-10-19 14:48:24.096 CEST [353469M] DEBUG_GC: New cache entry: domainID 1324, clientID 7, query_type 1 (ID 3578)
2024-10-19 14:48:24.100 CEST [353469M] DEBUG_GC: New cache entry: domainID 1394, clientID 7, query_type 16 (ID 3579)
2024-10-19 14:48:24.102 CEST [353469M] DEBUG_GC: New cache entry: domainID 1395, clientID 7, query_type 16 (ID 3580)
2024-10-19 14:48:24.103 CEST [353469M] DEBUG_GC: New cache entry: domainID 1324, clientID 7, query_type 16 (ID 3581)
2024-10-19 14:48:24.105 CEST [353469M] DEBUG_GC: New cache entry: domainID 1394, clientID 8, query_type 1 (ID 3582)
2024-10-19 14:48:24.106 CEST [353469M] DEBUG_GC: New cache entry: domainID 1395, clientID 8, query_type 1 (ID 3583)
2024-10-19 14:48:24.108 CEST [353469M] DEBUG_GC: New cache entry: domainID 1324, clientID 8, query_type 1 (ID 3584)
2024-10-19 14:48:24.112 CEST [353469M] DEBUG_GC: New cache entry: domainID 1394, clientID 8, query_type 16 (ID 3585)
2024-10-19 14:48:24.113 CEST [353469M] DEBUG_GC: New cache entry: domainID 1395, clientID 8, query_type 16 (ID 3586)
2024-10-19 14:48:24.115 CEST [353469M] DEBUG_GC: New cache entry: domainID 1324, clientID 8, query_type 16 (ID 3587)
2024-10-19 14:49:00.472 CEST [353469M] DEBUG_ANY: dnsmasq received signal 17
2024-10-19 14:49:00.505 CEST [353469M] DEBUG_ANY: dnsmasq received signal 17
2024-10-19 14:49:00.617 CEST [353469/T353546] DEBUG_GC: GC starting, mintime: 2024-10-18 14:50:00 CEST (1729255800), counters->queries = 22114
2024-10-19 14:49:00.636 CEST [353469/T353546] DEBUG_GC: delete_old_queries_from_db(): Deleted 44 (44) queries, new number of queries in memory: 22070
...
2024-10-19 14:49:00.663 CEST [353469/T353546] DEBUG_GC: Recycling domain gateway.prod.eu-west-1.forester.a2z.com (ID 1395, lastQuery at 0.000)
2024-10-19 14:49:00.663 CEST [353469/T353546] DEBUG_GC: Recycling cache entry with ID 1216
....
2024-10-19 14:49:00.673 CEST [353469/T353546] DEBUG_GC: Recycling cache entry with ID 3577
2024-10-19 14:49:00.673 CEST [353469/T353546] DEBUG_GC: Recycling cache entry with ID 3578
2024-10-19 14:49:00.673 CEST [353469/T353546] DEBUG_GC: Recycling cache entry with ID 3579
2024-10-19 14:49:00.674 CEST [353469/T353546] DEBUG_GC: Recycling cache entry with ID 3580
2024-10-19 14:49:00.674 CEST [353469/T353546] DEBUG_GC: Recycling cache entry with ID 3582
2024-10-19 14:49:00.674 CEST [353469/T353546] DEBUG_GC: Recycling cache entry with ID 3583
2024-10-19 14:49:00.674 CEST [353469/T353546] DEBUG_GC: Recycling cache entry with ID 3584
2024-10-19 14:49:00.674 CEST [353469/T353546] DEBUG_GC: Recycling cache entry with ID 3585
2024-10-19 14:49:00.674 CEST [353469/T353546] DEBUG_GC: Recycling cache entry with ID 3586
2024-10-19 14:49:00.678 CEST [353469/T353546] DEBUG_GC: 502/512 clients, 1710/3200 domains and 3683/7168 cache records are free
2024-10-19 14:49:00.679 CEST [353469/T353546] DEBUG_GC: Recycled additional 0 clients, 25 domains, and 103 cache records (scanned 22070 queries)
2024-10-19 14:49:00.679 CEST [353469/T353546] DEBUG_GC: GC removed 44 queries (took 62.57 ms)

```

**Comment by DL6ER** [comment]
> Is it expected to recycle the domains that fast?

Usually no but maybe yes. All domain **not used by any known query** are recycled every ten minutes. It was just coincidence that this domain happened to be added only a few seconds before.

The task is to find out why it isn't used by any query - maybe it was some intermediate domain found in CNAME inspection? Maybe, we really shouldn't recycle such domains as fast as they may re-appear a bit later. I could imagine changing the recycling condition from (pseudo code) `! used_by_query` to `! used_by_query && age > 24h`

For further investigation, I'd be ideal if you could provide some context for this domain from `pihole.log` (`FTL.log` with `debug.queries = true` would be even better but may not have been used at that time).

**Comment by yubiuser** [comment]
No FTl.log with `debug.queries` available. But it seems the domain is part of CNAME

```
nanopi@nanopi:~$ cat /var/log/pihole/pihole.log | grep -A 10 -B 5 gateway.prod.eu-west-1.forester.a2z.com
Oct 19 14:48:24 dnsmasq[353469]: forwarded fls-eu.amazon.de to 8.8.8.8
Oct 19 14:48:24 dnsmasq[353469]: query[HTTPS] fls-eu.amazon.de from 10.0.1.99
Oct 19 14:48:24 dnsmasq[353469]: forwarded fls-eu.amazon.de to 8.8.8.8
Oct 19 14:48:24 dnsmasq[353469]: reply fls-eu.amazon.de is <CNAME>
Oct 19 14:48:24 dnsmasq[353469]: reply fls-eu.amazon.com is <CNAME>
Oct 19 14:48:24 dnsmasq[353469]: reply gateway.prod.eu-west-1.forester.a2z.com is <CNAME>
Oct 19 14:48:24 dnsmasq[353469]: reply endpoint.prod.eu-west-1.forester.a2z.com is 52.211.180.158
Oct 19 14:48:24 dnsmasq[353469]: reply endpoint.prod.eu-west-1.forester.a2z.com is 99.80.70.202
Oct 19 14:48:24 dnsmasq[353469]: reply endpoint.prod.eu-west-1.forester.a2z.com is 52.51.149.79
Oct 19 14:48:24 dnsmasq[353469]: reply endpoint.prod.eu-west-1.forester.a2z.com is 52.210.162.248
Oct 19 14:48:24 dnsmasq[353469]: reply endpoint.prod.eu-west-1.forester.a2z.com is 34.251.164.146
Oct 19 14:48:24 dnsmasq[353469]: reply endpoint.prod.eu-west-1.forester.a2z.com is 52.17.171.122
Oct 19 14:48:24 dnsmasq[353469]: reply endpoint.prod.eu-west-1.forester.a2z.com is 54.76.34.181
Oct 19 14:48:24 dnsmasq[353469]: reply endpoint.prod.eu-west-1.forester.a2z.com is 52.17.159.245
Oct 19 14:48:24 dnsmasq[353469]: reply fls-eu.amazon.de is <CNAME>
Oct 19 14:48:24 dnsmasq[353469]: reply fls-eu.amazon.com is <CNAME>
Oct 19 14:48:24 dnsmasq[353469]: reply gateway.prod.eu-west-1.forester.a2z.com is <CNAME>
Oct 19 14:48:24 dnsmasq[353469]: reply endpoint.prod.eu-west-1.forester.a2z.com is NODATA
Oct 19 14:48:24 dnsmasq[353469]: reply fls-eu.amazon.de is <CNAME>
Oct 19 14:48:24 dnsmasq[353469]: reply fls-eu.amazon.com is <CNAME>
Oct 19 14:48:24 dnsmasq[353469]: reply gateway.prod.eu-west-1.forester.a2z.com is <CNAME>
Oct 19 14:48:24 dnsmasq[353469]: reply endpoint.prod.eu-west-1.forester.a2z.com is 34.241.70.112
Oct 19 14:48:24 dnsmasq[353469]: reply endpoint.prod.eu-west-1.forester.a2z.com is 54.171.170.141
Oct 19 14:48:24 dnsmasq[353469]: reply endpoint.prod.eu-west-1.forester.a2z.com is 34.249.53.81
Oct 19 14:48:24 dnsmasq[353469]: reply endpoint.prod.eu-west-1.forester.a2z.com is 52.212.71.64
Oct 19 14:48:24 dnsmasq[353469]: reply endpoint.prod.eu-west-1.forester.a2z.com is 108.129.4.132
Oct 19 14:48:24 dnsmasq[353469]: reply endpoint.prod.eu-west-1.forester.a2z.com is 54.170.54.143
Oct 19 14:48:24 dnsmasq[353469]: reply endpoint.prod.eu-west-1.forester.a2z.com is 52.19.180.243
Oct 19 14:48:24 dnsmasq[353469]: reply endpoint.prod.eu-west-1.forester.a2z.com is 54.75.200.212
Oct 19 14:48:24 dnsmasq[353469]: reply fls-eu.amazon.de is <CNAME>
Oct 19 14:48:24 dnsmasq[353469]: reply fls-eu.amazon.com is <CNAME>
Oct 19 14:48:24 dnsmasq[353469]: reply gateway.prod.eu-west-1.forester.a2z.com is <CNAME>
Oct 19 14:48:24 dnsmasq[353469]: reply endpoint.prod.eu-west-1.forester.a2z.com is NODATA
Oct 19 14:48:24 dnsmasq[353469]: query[HTTPS] endpoint.prod.eu-west-1.forester.a2z.com from 10.0.1.32
Oct 19 14:48:24 dnsmasq[353469]: cached endpoint.prod.eu-west-1.forester.a2z.com is NODATA
Oct 19 14:48:24 dnsmasq[353469]: query[HTTPS] endpoint.prod.eu-west-1.forester.a2z.com from 10.0.1.99
Oct 19 14:48:24 dnsmasq[353469]: cached endpoint.prod.eu-west-1.forester.a2z.com is NODATA
```

**Comment by DL6ER** [comment]
> Here is an other one, without being CNAME
> carmel.izzysoft.de

Well, it may just have been the last element of the CNAME chain. It may still not be what was requested. But I'm already convinced, commit incoming!

**Comment by yubiuser** [comment]
Huh,  see `gateway.kleinanzeigen.de.edgekey.net`

```
    Hash is 3f4bc0e2 (Latest: 3f4bc0e2)
nanopi@nanopi:~$ tail -f /var/log/pihole/FTL.log 
2024-10-19 16:37:37.289 CEST [363456M] DEBUG_GC: New domain: gateway.kleinanzeigen.de.edgekey.net (ID 1438)
2024-10-19 16:37:37.291 CEST [363456M] DEBUG_GC: New cache entry: domainID 1438, clientID 5, query_type 1 (ID 3367)
2024-10-19 16:37:37.293 CEST [363456M] DEBUG_GC: New cache entry: domainID 81, clientID 5, query_type 1 (ID 3368)
2024-10-19 16:37:37.538 CEST [363456M] DEBUG_GC: New cache entry: domainID 1425, clientID 5, query_type 1 (ID 3369)
2024-10-19 16:37:44.106 CEST [363456M] DEBUG_GC: New domain: prod-ws-2-2053862477.eu-central-1.elb.amazonaws.com (ID 1439)
2024-10-19 16:37:44.108 CEST [363456M] DEBUG_GC: New cache entry: domainID 1439, clientID 5, query_type 1 (ID 3370)
2024-10-19 16:38:00.483 CEST [363456M] DEBUG_ANY: dnsmasq received signal 17
2024-10-19 16:38:00.497 CEST [363456M] DEBUG_ANY: dnsmasq received signal 17
2024-10-19 16:38:07.717 CEST [363456M] DEBUG_GC: New cache entry: domainID 72, clientID 1, query_type 1 (ID 3371)
2024-10-19 16:38:08.214 CEST [363456M] DEBUG_GC: New cache entry: domainID 107, clientID 1, query_type 1 (ID 3372)
2024-10-19 16:38:26.411 CEST [363456M] DEBUG_GC: New domain: discourse.pi-hole.net (ID 1440)
2024-10-19 16:38:26.414 CEST [363456M] DEBUG_GC: New cache entry: domainID 1440, clientID 1, query_type 1 (ID 3373)
2024-10-19 16:38:26.420 CEST [363456M] DEBUG_GC: New cache entry: domainID 1440, clientID 1, query_type 16 (ID 3374)
2024-10-19 16:38:26.425 CEST [363456M] DEBUG_GC: New cache entry: domainID 1440, clientID 2, query_type 1 (ID 3375)
2024-10-19 16:38:26.430 CEST [363456M] DEBUG_GC: New cache entry: domainID 1440, clientID 2, query_type 16 (ID 3376)
2024-10-19 16:38:30.206 CEST [363456M] DEBUG_GC: New cache entry: domainID 673, clientID 1, query_type 1 (ID 3377)
2024-10-19 16:38:30.898 CEST [363456M] DEBUG_GC: New cache entry: domainID 675, clientID 1, query_type 1 (ID 3378)
2024-10-19 16:38:51.119 CEST [363456M] DEBUG_GC: New cache entry: domainID 1424, clientID 0, query_type 1 (ID 3379)
2024-10-19 16:39:00.520 CEST [363456/T363619] DEBUG_GC: GC starting, mintime: 2024-10-18 16:40:00 CEST (1729262400), counters->queries = 19997
2024-10-19 16:39:00.535 CEST [363456/T363619] DEBUG_GC: delete_old_queries_from_db(): Deleted 32 (32) queries, new number of queries in memory: 19965
2024-10-19 16:39:00.554 CEST [363456/T363619] DEBUG_GC: Recycling domain jobs.zeit.de (ID 7, lastQuery at 1729262281.829)
2024-10-19 16:39:00.556 CEST [363456/T363619] DEBUG_GC: Recycling domain ac88393aca5853df7.dualstack.awsglobalaccelerator.com (ID 1437, lastQuery at 0.000)
2024-10-19 16:39:00.558 CEST [363456/T363619] DEBUG_GC: Recycling domain gateway.kleinanzeigen.de.edgekey.net (ID 1438, lastQuery at 0.000)
2024-10-19 16:39:00.559 CEST [363456/T363619] DEBUG_GC: Recycling domain prod-ws-2-2053862477.eu-central-1.elb.amazonaws.com (ID 1439, lastQuery at 0.000)
2024-10-19 16:39:00.561 CEST [363456/T363619] DEBUG_GC: Recycling cache entry with ID 14
2024-10-19 16:39:00.563 CEST [363456/T363619] DEBUG_GC: Recycling cache entry with ID 15
2024-10-19 16:39:00.565 CEST [363456/T363619] DEBUG_GC: Recycling cache entry with ID 3352
2024-10-19 16:39:00.567 CEST [363456/T363619] DEBUG_GC: Recycling cache entry with ID 3353
2024-10-19 16:39:00.569 CEST [363456/T363619] DEBUG_GC: Recycling cache entry with ID 3366
2024-10-19 16:39:00.570 CEST [363456/T363619] DEBUG_GC: Recycling cache entry with ID 3367
2024-10-19 16:39:00.572 CEST [363456/T363619] DEBUG_GC: Recycling cache entry with ID 3368
2024-10-19 16:39:00.574 CEST [363456/T363619] DEBUG_GC: Recycling cache entry with ID 3369
2024-10-19 16:39:00.576 CEST [363456/T363619] DEBUG_GC: Recycling cache entry with ID 3370
2024-10-19 16:39:00.578 CEST [363456/T363619] DEBUG_GC: Recycling cache entry with ID 3371
2024-10-19 16:39:00.580 CEST [363456/T363619] DEBUG_GC: Recycling cache entry with ID 3372
2024-10-19 16:39:00.582 CEST [363456/T363619] DEBUG_GC: Recycling cache entry with ID 3377
2024-10-19 16:39:00.584 CEST [363456/T363619] DEBUG_GC: Recycling cache entry with ID 3378
2024-10-19 16:39:00.585 CEST [363456/T363619] DEBUG_GC: Recycling cache entry with ID 3379
2024-10-19 16:39:00.586 CEST [363456M] DEBUG_ANY: dnsmasq received signal 17
2024-10-19 16:39:00.588 CEST [363456/T363619] DEBUG_GC: 502/512 clients, 99/1536 domains and 218/3584 cache records are free
2024-10-19 16:39:00.590 CEST [363456/T363619] DEBUG_GC: Recycled additional 0 clients, 4 domains, and 14 cache records (scanned 19965 queries)
```

**Comment by DL6ER** [comment]
Yeah, it's exactly the same, I guess? Hard to say without `pihole.log`. Anyway, if you restart from the very last commits in this PR, only domains being unused for > 24 hours should be recycled.

**Comment by yubiuser** [comment]
See the first line... This is the latest hash

**Comment by DL6ER** [comment]
Ah, yes, see us two - talking here about CNAME domains and I did forget to add the `lastQuery` updating for *exactly* this scenario. :face_with_peeking_eye: 

Should now be fixed. The `lastQuery at 0.000` is what is wrong here - there should always be a timestamp set here.

**Comment by yubiuser** [comment]
```
2024-10-19 22:33:05.275 CEST [368102M] INFO: Received: RT42 (42 -> 7)
2024-10-19 22:33:06.123 CEST [368102/T368105] INFO: Searching for hash collisions in clients lookup table
2024-10-19 22:33:06.123 CEST [368102/T368105] INFO: Found 0 hash collisions in clients lookup table (scanned 10 elements)
2024-10-19 22:33:06.124 CEST [368102/T368105] INFO: Searching for hash collisions in domains lookup table
2024-10-19 22:33:06.124 CEST [368102/T368105] INFO: Found 0 hash collisions in domains lookup table (scanned 1483 elements)
2024-10-19 22:33:06.124 CEST [368102/T368105] INFO: Searching for hash collisions in DNS cache lookup table
2024-10-19 22:33:06.124 CEST [368102/T368105] INFO: Found 0 hash collisions in DNS cache lookup table (scanned 3455 elements)

```


> We can discuss if we leave it in or if we want to remove it before finally merging this PR

We should keep it. It could be handy at some point. But please amend
https://deploy-preview-338--pihole-docs.netlify.app/ftldns/signals/

**Review by yubiuser — APPROVED** [review]
Looks really nice now. 

Please address

> We should keep it. It could be handy at some point. But please amend
https://deploy-preview-338--pihole-docs.netlify.app/ftldns/signals/

**Comment by DL6ER** [comment]
https://github.com/pi-hole/docs/commit/762f4c1516777b95f7cccb354d4ab1906918ec52
